ProjFS.Mac/PrjFSKext/public/Message.h

```diff
@@ -47,25 +47,38 @@ struct MessageHeader
     int32_t             pid;
     char                procname[MAXCOMLEN + 1];
 
-    // Size of the flexible-length, nul-terminated path following the message body, including the nul character.
-    uint16_t            pathSizeBytes;
+    // Size of the flexible-length, arbitrary data following the message body.
+    // For most message types this can either be empty or a nul-terminated
+    // string, where dataSizeBytes includes the termination byte.
+    // For the FileHardLinkCreated and *Renamed messages, this is always a
+    // chained string buffer containing 2 path strings.
+    uint16_t            dataSizeBytes;
 };
 
 // Description of a decomposed, in-memory message header plus variable length string field
 struct Message
 {
     const MessageHeader* messageHeader;
-    const char*    path;
+    const void*          data;
+    uint16_t             dataSizeBytes;
```

Comment:
>>This opens us up to easily pass more info in the future. I wanted to double check the approach here. Would we want to try to have actual fields for the extra members or is an array the right approach. Downside to the array is you need to know what each spot represents. I'm honestly not sure the best approach just wanted to call it out. @wilbaker

>In the case of strings/paths, if you were to stick it all into a struct, you'd have to send the maximum possible size every time, rather than just the length of the actual strings.

>I'd say put small (non-string/non-binary-blob) fields in MessageHeader and put anything variable-length in the general data area. For now we only need (a variable number of) strings, but in principle you could send other data down there too.

Moving this conversation here to make threaded conversation a bit easier.

I like @jeschu1's idea of having explicit fields for each path, and I also agree with @pmj that variable length data should go in the general data area.

Can we combine both of these ideas and go with something like this?

```
struct MessageHeader
{
    ...
    // Sizes of the flexible-length, nul-terminated paths following the message body, including the nul character.	   
    uint16_t            pathSizeBytes;
    uint16_t            fromPathSizeBytes;
```

```
struct Message
{
    const MessageHeader* messageHeader;
    const char*    path;
    const char*    fromPath;
}
```

We would be losing some flexibility, but in exchange we'd gain:

- More explicit validation (e.g. `assert(vnodePath != nullptr` rather than `assert(dataStringCount > 0`)
- Immediate access to all message fields (i.e. no need to parse `data` to split strings)
- Easier to view path values when debugging

@jeschu1, @pmj, what do you think?

Reply:
@jeschu1 I like the idea of keeping it more explicit for now.  I *think* this should make for a slightly easier PR.  @pmj thoughts on trying approach @wilbaker suggests above for comparison.

Note the first commit in this batch is approved in a separate PR, so the number of changes here will drop.

Reply:
Yep, we can do that. I'm going to experiment with putting the sizes in a `struct` inside a `union` to share space with an array, so that we don't have to write special-case code for each field, and can just loop over them with an index when encoding/decoding. (And yes, this is valid C & C++ as long as the types match.) So that would be:

```
struct MessageHeader
{
   …
   union
   {
      uint16_t sizesBytes[2];
      struct
      {
          uint16_t            pathSizeBytes;
          uint16_t            fromPathSizeBytes;
      };
   };
};
```

"Experiment" because I'm not 100% sure yet if the code will be simpler this way or not. If not, I'll go with the "raw" approach as suggested.

Reply:
I've reworked the message format change commit as above now. (48ac789) The `union` with the array does make the code less repetitive, obviously this will compound if we go beyond 2 variable-length fields.

Reply:
@pmj I totally see how that works!  In my head I'm thinking the extra field may be simplier code-wise.  Any chance you could know that up for comparison?

Reply:
I've just rebased the PR on latest master and (hopefully) fixed the conflicts with the vnode cache changes. I've also pushed a series of commits that remove the union and just use the struct fields. I don't like this code in either of the 2 iterations. I think if we really must remove the union, I'd prefer to stick with the array part of it and use named (enum) array indices instead.